index.js

```diff
@@ -0,0 +1,106 @@
+const { default: axios } = require('axios');
+const { Handler } = require('htmlmetaparser');
+const { Parser } = require('htmlparser2');
+
+/**
+ * This is a recursive function that returns an array of dataset site URLs.
+ * If the URL supplied is a data catalog collection, it takes all the part collections in hasPart and crawls them.
+ * If the URL supplied is a data catalog, it takes the dataset array and flattens them. 
+ * 
+ * @param {string} dataCatalogUrl
+ * @returns {Promise<string[]>}
+ */
+async function recursivelyGetCatalogsToGetDatasetUrls(dataCatalogUrl) {
+  let catalog;
+  try {
+    catalog = (await axios.get(dataCatalogUrl, {timeout: 5000})).data;
+  } catch (error) {
+    console.error(`Error getting catalog or catalog collection, url: ${dataCatalogUrl}`)
+    return [];
+  }
+
+  // If catalog has hasPart, the part catalog must be fetched and the datasets got from the part catalog
+  // The part catalog could have a part catalog within in, which is why this function must be recursive.
+  if (catalog.hasPart) {
+    const datasetArray = await Promise.all(catalog.hasPart.map(partCatalogUrl => recursivelyGetCatalogsToGetDatasetUrls(partCatalogUrl)));
+    return [].concat(...datasetArray);
+  }
+
+  // If the catalog has dataset, it does not have ant further part catalogs and the datasets can be got from them
+  if (catalog.dataset) {
+    return catalog.dataset;
+  }
+
+  // If the catalog has neither hasPart or dataset, return [] as it does not have the information we want
+  return [];
+}
+
+/**
+ * This function extracts JSONLD metadata from dataset HTML
+ * 
+ * @param {string} url 
+ * @param {string} html 
+ */
+function extractJSONLDfromHTML(url, html) {
+  let jsonld = null;
+
+  const handler = new Handler(
+    (err, result) => {
+      if (!err && typeof result === 'object') {
+        const jsonldArray = result.jsonld;
+        // Use the first JSON-LD block on the page
+        if (Array.isArray(jsonldArray) && jsonldArray.length > 0) {
+          [jsonld] = jsonldArray;
+        }
+      }
+    },
+    {
+      url, // The HTML pages URL is used to resolve relative URLs. TODO: Remove this
+    },
+  );
+
+  // Create a HTML parser with the handler.
+  const parser = new Parser(handler, {
+    decodeEntities: true,
+  });
+  parser.write(html);
+  parser.done();
+
+  return jsonld;
+}
+
+/**
+ * This function recursively crawls through a data catalog, fetches datasets, and extracts JSONLD
+ * from dataset HTML.
+ * This combines recursivelyGetCatalogsToGetDatasetUrls() and extractJSONLDfromHTML()
+ * 
+ * @param {string} dataCatalogUrl
+ */
+async function extractJSONLDfromDatasetsFromDataCatalog(dataCatalogUrl) {
```

Comment:
I'm wondering whether this might be clearer, in that it doesn't require a deep understanding of how the data catalogs are parsed to understand what the function does?

```suggestion
async function getAllDatasets(dataCatalogUrl) {
```